flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/ResettableExternalBuffer.java

```diff
@@ -0,0 +1,699 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.util;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.io.disk.RandomAccessInputView;
+import org.apache.flink.runtime.io.disk.SimpleCollectingOutputView;
+import org.apache.flink.runtime.io.disk.iomanager.BlockChannelReader;
+import org.apache.flink.runtime.io.disk.iomanager.BlockChannelWriter;
+import org.apache.flink.runtime.io.disk.iomanager.ChannelReaderInputView;
+import org.apache.flink.runtime.io.disk.iomanager.FileIOChannel;
+import org.apache.flink.runtime.io.disk.iomanager.HeaderlessChannelReaderInputView;
+import org.apache.flink.runtime.io.disk.iomanager.IOManager;
+import org.apache.flink.runtime.memory.ListMemorySegmentSource;
+import org.apache.flink.runtime.memory.MemoryManager;
+import org.apache.flink.table.dataformat.BaseRow;
+import org.apache.flink.table.dataformat.BinaryRow;
+import org.apache.flink.table.runtime.io.BinaryRowChannelInputViewIterator;
+import org.apache.flink.table.runtime.io.ChannelWithMeta;
+import org.apache.flink.table.typeutils.AbstractRowSerializer;
+import org.apache.flink.table.typeutils.BinaryRowSerializer;
+import org.apache.flink.util.MutableObjectIterator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.EOFException;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * A resettable external buffer for binary row. It stores records in memory and spill to disk
+ * when memory is not enough. When the spill is completed, the records are written to memory again.
+ * The returned iterator reads the data in write order (read spilled records first).
+ *
+ * <p>1.It supports infinite length.
+ * 2.It can open multiple Iterators.
+ * 3.It support new iterator with beginRow.
+ */
+public class ResettableExternalBuffer implements Closeable {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ResettableExternalBuffer.class);
+
+	// We will only read one spilled file at the same time.
+	private static final int READ_BUFFER = 2;
+
+	private final MemoryManager memoryManager;
+	private final IOManager ioManager;
+	private final List<MemorySegment> memory;
+	private final BinaryRowSerializer binaryRowSerializer;
+	private final InMemoryBuffer inMemoryBuffer;
+	private long spillSize;
+
+	// The size of each segment
+	private int segmentSize;
+
+	// The length of each row, if each row is of fixed length
+	private long rowFixedPartLength;
+	// If each row is of fixed length
+	private boolean isRowAllInFixedPart;
```

Comment:
isFixedLengthRow

Reply:
I think isFixedLengthRow is misunderstandable, The implication here is that optimization can only be done when row all in fixed part.

Reply:
what's the differences between "row all in fixed part" and "the row is fixed length"?

Reply:
For example, now a field is Decimal, and its precision is greater than 18, which means that it needs to be stored in a variable length part, and the length in variable length part must be 18.
In this case, it is fixed length, but not entirely in the fixed length part.
ResettableExternalBuffer's rowLength is get from `binaryRowSerializer.getSerializedRowFixedPartLength`, it mean that the rowLength is the length of fixed length part, not total size maybe.

Reply:
But in this case, we can still use the improvement we currently have once the row is fixed length, right? And it's really hard for a new developer to get the point like this...

Reply:
We can, but we still not do it..